src/main/scala/calc/Typechecker.scala

```diff
@@ -0,0 +1,142 @@
+package calc
+
+import org.scalajs.core.ir.{Types => irtpe, Position}
+
+/** A simple typechecker.
+ *
+ * Checks whether the types match or are otherwise appropriate (e.g. in the if's condition).
+ * Annotates AST with types, so that they can be used during compilation.
+ *
+  * TODO group errors and fail as late as possible, possibly without an exception (requires some type-work)
+ */
+
+object Typechecker {
+
+  class TypecheckerException(msg: String) extends Exception(msg)
+
+  type TypeScope = Map[String, TreeType]
+
+  implicit def getScalaJSType(tp: TreeType): irtpe.Type = tp match {
+    case DoubleType      => irtpe.DoubleType
+    case ParamType       => irtpe.AnyType
+    case FunctionType(_) => irtpe.AnyType
+    case NoType          => irtpe.NoType
+  }
+
+  def getType(tree: Tree, scope: TypeScope)
+             (implicit pos: Position): Tree =  tree match {
+    case Literal(x, _)    => Literal(x, DoubleType)
+    case b:      BinaryOp => checkBinop(b, scope)
+    case id:     Ident    => checkIdent(id, scope)
+    case let:    Let      => checkLet(let, scope)
+    case ifExpr: If       => checkIf(ifExpr, scope)
+    case call:   Call     => checkCall(call, scope)
+    case cl:     Closure  => checkClosure(cl, scope)
+  }
+
+  def typecheck(tree: Tree)(implicit pos: Position): Tree = {
+    getType(tree, Map.empty[String, TreeType])
+  }
+
+  private def checkBinop(binop: BinaryOp, scope: TypeScope)
+                        (implicit pos: Position): Tree = {
+    val lhsTyped = getType(binop.lhs, scope)
+    val rhsTyped = getType(binop.rhs, scope)
+    val lType = lhsTyped.tp
+    val rType = rhsTyped.tp
+
+    if (!lType.isFunction && !rType.isFunction)
+      BinaryOp(binop.op, lhsTyped, rhsTyped, lType)
+    else
+      fail(binop.pos, s"Incompatible operand types for ${binop.op}: $lType, $rType")
+  }
+
+  private def checkIdent(ident: Ident, scope: TypeScope)
+                        (implicit pos: Position): Tree =
+    scope.get(ident.name) match {
+      case Some(t) => Ident(ident.name, t)
+      case None    => fail(pos, s"Not in scope: ${ident.name}")
+    }
+
+  private def checkIf(ifExpr: If, scope: TypeScope)
+                     (implicit pos: Position): Tree = ifExpr match {
+    case If(lit@Literal(_, _), thenp, elsep, _) =>
+      val thenTyped = getType(thenp, scope)
+      val elseTyped = getType(elsep, scope)
+      val tType = thenTyped.tp
+      val eType = elseTyped.tp
+
+      if (tType != eType)
+        fail(ifExpr.pos,
+          s"Both branches of if should evaluate to the same type. " +
+          s"Got: $tType and $eType instead.")
+
+      If(getType(lit, scope), thenTyped, elseTyped, tType)
+
+    case _ => fail(ifExpr.pos, s"If condition should be a number")
```

Comment:
Er ... what about `let x = 5 in (if (x) 1 else 2)`?
Or `let f = (fun(x) = { x + 1 }) in if (f(0)) 4 else 5`?